diff-containers/test/Test/Data/Map/Diff/Strict.hs

```diff
@@ -174,3 +273,78 @@ instance Arbitrary v => Arbitrary (DiffEntry v) where
+newtype PositiveDiff k v = PositiveDiff { getPositiveDiff :: Diff k v }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance (Ord k, Arbitrary k, Arbitrary v)
+      => Arbitrary (PositiveDiff k v) where
+  arbitrary               = PositiveDiff <$> arbitrary `suchThat` isPositive
+  shrink (PositiveDiff d) = PositiveDiff <$> shrink d
+
+-- | A @'DiffHistory'@ for which @'isPositiveDiffHistory'@ holds.
+newtype PositiveDiffHistory v = PositiveDiffHistory {
+    getPositiveDiffHistory :: DiffHistory v
+  }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance Arbitrary v => Arbitrary (PositiveDiffHistory v) where
+  arbitrary = PositiveDiffHistory <$> arbitrary `suchThat` isPositiveDiffHistory
+  shrink (PositiveDiffHistory h) = PositiveDiffHistory <$> shrink h
+
+{------------------------------------------------------------------------------
+  Modifiers: normality
+------------------------------------------------------------------------------}
+
+-- | A @'Diff'@ for which @'isNormal'@ holds.
+newtype NormalDiff k v = NormalDiff { getNormalDiff :: Diff k v }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance (Ord k, Eq v, Arbitrary k, Arbitrary v)
+      => Arbitrary (NormalDiff k v) where
+  arbitrary             = NormalDiff <$> arbitrary `suchThat` isNormal
+  shrink (NormalDiff d) = NormalDiff <$> shrink d
+
+-- | A @'DiffHistory'@ for which @'isNormalDiffHistory'@ holds.
+newtype NormalDiffHistory v = NormalDiffHistory {
+    getNormalDiffHistory :: DiffHistory v
+  }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance (Arbitrary v, Eq v) => Arbitrary (NormalDiffHistory v) where
+  arbitrary = NormalDiffHistory <$> arbitrary `suchThat` isNormalDiffHistory
+  shrink (NormalDiffHistory h) = NormalDiffHistory <$> shrink h
+
+{------------------------------------------------------------------------------
+  Modifiers: positivity and normality
+------------------------------------------------------------------------------}
+
+-- | A @'Diff'@ for which @'isNormal'@ and @'isPositive'@ hold.
+newtype NPDiff k v = NPDiff { getNPDiff :: Diff k v }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance (Ord k, Eq v, Arbitrary k, Arbitrary v)
+      => Arbitrary (NPDiff k v) where
+  arbitrary             = NPDiff <$> arbitrary `suchThat` (\d -> isNormal d && isPositive d)
+  shrink (NPDiff d) = NPDiff <$> shrink d
+
+-- | A @'DiffHistory'@ for which @'isNormalDiffHistory'@ and @'isPositiveDiffHistory'@ hold.
+newtype NPDiffHistory v = NPDiffHistory {
+    getNPDiffHistory :: DiffHistory v
+  }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance (Arbitrary v, Eq v) => Arbitrary (NPDiffHistory v) where
+  arbitrary = NPDiffHistory <$> arbitrary `suchThat` (\d -> isNormalDiffHistory d && isPositiveDiffHistory d)
```

Comment:
Same here: we should probably check that the shrinking function preserves the invariant.